python_bindings/caramel/__init__.py

```diff
@@ -0,0 +1,75 @@
+from ._caramel import *
+
+
+def CSF(keys, values, max_to_validate=None):
+    """
+    Constructs a CSF, automatically inferring the correct CSF backend.
+
+    Arguments:
+        keys: List of hashable keys.
+        values: List of values to use in the CSF.
+        max_to_validate: If provided, only the first "max_to_validate" values
+            will be checked when inferring the correct CSF backend.
+    
+    Returns:
+        A CSF containing the desired key-value mapping.
+
+    Raises:
+        ValueError if the keys and values cannot be used to construct a CSF.
+    """
+    if not len(keys):
+        raise ValueError("Keys must be non-empty but found length 0.")
+    if not len(values):
+        raise ValueError("Values must be non-empty but found length 0.")
+    if len(keys) != len(values):
+        raise ValueError("Keys and values must have the same length.")
+    if not isinstance(keys[0], (str, bytes)):
```

Comment:
should we check the first max_to_validate values?

Reply:
Good question - the max_to_validate is for type inference (not error handling). That is, if the first max_to_validate values are length-12 strings, we assume that that all of the values are length-12 strings and we dispatch to CSFChar12. CSFChar12 will throw an exception (inside the pybind11 binding casts) if this isn't the case. So it really just means that "backend inference is based on the first max_to_validate." Motivation here is that if you have some 100M-sized CSF, maybe it's sufficient to do type inference on the first 100k (checking 100M values in Python could be painful / slow).

This can definitely be documented better, will fix. In particular, we don't validate all the keys or values since the construction process throws an exception upon encountering an invalid key/value.